fms_extras/models/paged_llama.py

```diff
@@ -0,0 +1,844 @@
+            position_ids: torch.Tensor, optional
+                the optional position ids to be used with rotary embeddings. If None, will just be positions ids
+                starting at position 0 for each sequence in the batch
+            cache_data: PagedAttentionCacheData, optional
+                the optional cache data used in paged attention. If None is given, SDPA will always be used
+            use_cache: bool
+                denotes whether a cache should be used. If True, the cache will be returned as well as the logits
+            only_last_token: bool
+                only return the last token from embedding output
+            attn_algorithm: str, optional
+                string to denote which attention algorithm to use
+            return_embeds: bool
+                If True, will return the embeddings, otherwise wil not return embeddings
+
+        Returns:
+        Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]
+            if use_cache and return_embeds is True => (logits, cache, embeds)
+            if use_cache is True => (logits, cache)
+            if return_embeds is True => (logits, embeds)
+            otherwise => logits
+        """
+        embeds, cache = self.headless_model(
+            x,
+            mask,
+            position_ids,
+            cache_data,
+            use_cache,
+            attn_algorithm,
+        )
+
+        if only_last_token:
+            embeds = embeds[:, -1, :]
+
+        preds = self.head(embeds, reverse=True)
+
+        out = [preds]
+        if use_cache:
+            out.append(cache)
+        if return_embeds:
+            out.append(embeds)
+        if len(out) == 1:
+            return out[0]
+        else:
+            return tuple(out)
+
+
+# Register common LLaMA variants with the model registration API
+
+# a micro llama model to use with a char-level tokenizer
+_micro_char_config = PagedLLaMAConfig(
+    emb_dim=192, nheads=4, nlayers=5, max_expected_seq_len=1024, src_vocab_size=256
+)
+
+_7b_config = PagedLLaMAConfig()
+_13b_config = PagedLLaMAConfig(emb_dim=5120, nheads=40, nlayers=40)
+# todo: add 35B config
+
+_70b_config = PagedLLaMAConfig(
+    emb_dim=8192,
+    multiple_of=4096,
+    nheads=64,
+    kvheads=8,
+    nlayers=80,
+    hidden_grow_factor=(1.3 * 8 / 3),
+)
+
+_architecture_name = "paged_llama"
+
+
+def _llama_factory_factory(config):
+    def factory(**kwargs):
+        return PagedLLaMA(config, **kwargs)
+
+    return factory
+
+
+models.register_model(
+    _architecture_name, "micro", _llama_factory_factory(_micro_char_config)
+)
+models.register_model(_architecture_name, "7b", _llama_factory_factory(_7b_config))
+models.register_model(_architecture_name, "13b", _llama_factory_factory(_13b_config))
+models.register_model(_architecture_name, "70b", _llama_factory_factory(_70b_config))
+
+
+def _rename_weights_to_fms(orig_sd):
+    replacements = [
```

Comment:
i won't block this PR on it, and I know there are some small differences, but I suspect there's something better we could be doing with these adapters to avoid so much repetition.